File: 10194/10194.cpp
#include<algorithm>
#include<vector>
#include<map>
#include<string>
#include<iostream>
using namespace std;

struct Team {
    string name;
    int pointsEarned;
    int wins;
    int losses;
    int ties;
    int goalsScored;
    int goalsAgainst;
    int gamesPlayed;
    Team(string name):
        name(name),
        pointsEarned(0),
        wins(0),
        losses(0),
        ties(0),
        goalsScored(0),
        goalsAgainst(0),
        gamesPlayed(0)
    {}
};

bool most_points(Team* t1, Team* t2){
    return t2->pointsEarned < t1->pointsEarned;
}

bool most_wins (Team* t1, Team* t2){
    return t2->wins < t1->wins;
}

bool most_goal_difference (Team* t1, Team* t2){
    return (t2->goalsScored - t2->goalsAgainst)
        < (t1->goalsScored - t1->goalsAgainst);
}

bool most_goals_scored (Team* t1, Team* t2){
    return t2->goalsScored < t1->goalsScored; 
}

bool least_games_played (Team* t1, Team* t2){
    return t1->gamesPlayed < t2->gamesPlayed; 
}

bool by_name (Team* t1, Team* t2){
    return t1->name < t2->name; 
}

int main(){
    string junk;

    int N;
    cin >> N;
    getline(cin, junk); // discard trailing newline

    for(int i = 0; i < N; i++){
        string tournamentName;
        getline(cin, tournamentName);

        //cerr << tournamentName << endl;

        map<string, Team*> teamMap;
        vector<Team*> teams;

        int T;
        cin >> T;
        getline(cin, junk);
        for(int j = 0; j < T; j++){
            string teamName;
            getline(cin, teamName);
            //cerr << '\t' << teamName << endl;

            Team* team = new Team(teamName);
            teams.push_back(team);
            teamMap.insert(make_pair(teamName, team));
        }

        int G;
        cin >> G;
        getline(cin, junk);
        for(int k = 0; k < G; k++){
            string game;
            getline(cin, game);
            int firstNum  = game.find("#");
            int lastNum  = game.rfind("#")+1;
            string team1Name = game.substr(0, firstNum);
            string team2Name = game.substr(lastNum);
            string points = game.substr(
                    firstNum+1, 
                    lastNum-firstNum);
            int at = points.find("@");
            int team1Points = stoi(points.substr(0, at));
            int team2Points = stoi(points.substr(at+1));

            //cerr << '\t' << team1Name << " vs " << team2Name << endl;

            Team* team1 = teamMap[team1Name];
            Team* team2 = teamMap[team2Name];

            (team1->goalsScored) += team1Points;
            (team2->goalsScored) += team2Points;

            (team1->goalsAgainst) += team2Points;
            (team2->goalsAgainst) += team1Points;

            if(team1Points > team2Points){
                (team1->wins)++;
                (team1->pointsEarned) += 3;
                (team2->losses)++;
            }
            else if(team2Points > team1Points){
                (team2->wins)++;
                (team2->pointsEarned) += 3;
                (team1->losses)++;
            }
            else {
                (team1->ties)++;
                (team2->ties)++;
                (team1->pointsEarned)++;
                (team2->pointsEarned)++;
            }

            (team1->gamesPlayed)++;
            (team2->gamesPlayed)++;
        }

        stable_sort(teams.begin(), teams.end(), by_name);
        stable_sort(teams.begin(), teams.end(), least_games_played);
        stable_sort(teams.begin(), teams.end(), most_goals_scored);
        stable_sort(teams.begin(), teams.end(), most_goal_difference);
        stable_sort(teams.begin(), teams.end(), most_wins);
        stable_sort(teams.begin(), teams.end(), most_points);

        cout << tournamentName << endl;
        int rank = 0;
        for(vector<Team*>::iterator it=teams.begin(); it != teams.end(); it++){
            // "%d) %s %dp, %dg, (%d-%d-%d), %dgd (%d-%d)\n"
            cout
                << ++rank << ") "
                << (*it)->name << " "
                << (*it)->pointsEarned << "p, "
                << (*it)->gamesPlayed << "g, "
                << "("
                << (*it)->wins << "-"
                << (*it)->ties << "-"
                << (*it)->losses
                << "), "
                << (*it)->goalsScored - (*it)->goalsAgainst << "gd "
                << "("
                << (*it)->goalsScored << "-"
                << (*it)->goalsAgainst
                << ")"
                << endl;
            delete (*it);
        }
        cout << endl;
    }

    return 0;
}

File: 10194/10194.py
#!/usr/bin/env python3
import sys

class Team:
    def __init__(self, name):
        self.name = name
        self.pointsEarned = 0
        self.wins = 0
        self.losses = 0
        self.ties = 0
        self.goalsScored = 0
        self.goalsAgainst = 0
        self.goalDifference = 0
        self.gamesPlayed = 0

    def __str__(self):
        return ('{0.name} '
                '{0.pointsEarned}p, '
                '{0.gamesPlayed}g, '
                '({0.wins}-{0.ties}-{0.losses}), '
                '{0.goalDifference}gd '
                '({0.goalsScored}-{0.goalsAgainst})').format(self)

N = int(next(sys.stdin))

for i in range(N):
    tournamentName = next(sys.stdin).strip()
    teamMap = {}
    teams = []

    T = int(next(sys.stdin))
    for j in range(T):
        teamName = next(sys.stdin).strip()
        team = Team(teamName)
        teamMap[teamName] = team
        teams.append(team)

    G = int(next(sys.stdin))
    for k in range(G):
        game = next(sys.stdin).strip()
        team1Info, team2Info = game.split("@")

        team1Name, team1Points = team1Info.split("#")
        team2Points, team2Name = team2Info.split("#")

        team1Points, team2Points = map(int, [team1Points, team2Points])

        team1 = teamMap[team1Name]
        team2 = teamMap[team2Name]

        team1.goalsScored += team1Points
        team2.goalsScored += team2Points

        team1.goalsAgainst += team2Points
        team2.goalsAgainst += team1Points

        team1.goalDifference = team1.goalsScored - team1.goalsAgainst
        team2.goalDifference = team2.goalsScored - team2.goalsAgainst

        if(team1Points > team2Points):
            team1.wins += 1
            team1.pointsEarned += 3
            team2.losses += 1
        elif(team2Points > team1Points):
            team2.wins += 1
            team2.pointsEarned += 3
            team1.losses += 1
        else :
            team1.ties += 1
            team2.ties += 1
            team1.pointsEarned += 1
            team2.pointsEarned += 1

        team1.gamesPlayed += 1
        team2.gamesPlayed += 1

    teams.sort(key=lambda t: t.name.lower())
    teams.sort(key=lambda t: t.gamesPlayed)
    teams.sort(key=lambda t: t.goalsScored, reverse=True)
    teams.sort(key=lambda t: t.goalDifference, reverse=True)
    teams.sort(key=lambda t: t.wins, reverse=True)
    teams.sort(key=lambda t: t.pointsEarned, reverse=True)

    print(tournamentName)
    for rank, t in enumerate(teams):
        print("{}) {}".format( rank+1, t ))
    print()

